modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py

```diff
@@ -846,52 +846,127 @@ def join(
         HdkOnNativeDataframe
             The new frame.
         """
-        how = JoinType(how)
+        check_join_supported(how)
         assert (
             left_on is not None and right_on is not None
         ), "Merge with unspecified 'left_on' or 'right_on' parameter is not supported in the engine"
         assert len(left_on) == len(
             right_on
         ), "'left_on' and 'right_on' lengths don't match"
 
-        for col in left_on:
-            assert col in self.columns, f"'left_on' references unknown column {col}"
-        for col in right_on:
-            assert col in other.columns, f"'right_on' references unknown column {col}"
-
-        new_columns = []
+        def validate(what, df, col_names):
+            cols = df.columns
+            new_col_names = col_names
+            for i, col in enumerate(col_names):
+                if col not in cols:
+                    new_name = None
+                    if df._index_cache is not None:
+                        for j, name in enumerate(df._index_cache):
+                            if name == col:
+                                new_name = f"__index__{j}_{name}"
+                                df = df._maybe_materialize_rowid()
+                                break
+                    if (new_name is None) and (df._index_cols is not None):
+                        for c in df._index_cols:
+                            if col == df._index_name(c):
+                                new_name = c
+                                break
```

Comment:
by the way, can we simplify this as follows?
```suggestion
                   if df._index_cols is not None:
                       for c in df._index_cols:
                           if col == df._index_name(c):
                               new_name = c
                               break
                    elif df._index_cache is not None:
                        new_name = f"__index__{0}_{name}"
                        df = df._maybe_materialize_rowid()
```
is there a case when we have a multi-index stored inside of `._index_cache` and `._index_cols is None`?

Reply:
Probably, there is no such a case. Thanks.